src/Config/MergeConfig.cs

```diff
@@ -0,0 +1,126 @@
+    {
+        /// <summary>
+        /// Parse the string and call the helper methods to merge based on their value kind, i.e. Objects or Array.
+        /// </summary>
+        public static string Merge(string originalJson, string overridingJson)
+        {
+            ArrayBufferWriter<byte> outputBuffer = new();
+
+            using (JsonDocument originalJsonDoc = JsonDocument.Parse(originalJson))
+            using (JsonDocument overridingJsonDoc = JsonDocument.Parse(overridingJson))
+            using (Utf8JsonWriter jsonWriter = new(outputBuffer, new JsonWriterOptions { Indented = true }))
+            {
+                JsonElement root1 = originalJsonDoc.RootElement;
+                JsonElement root2 = overridingJsonDoc.RootElement;
+
+                if (root1.ValueKind != JsonValueKind.Array && root1.ValueKind != JsonValueKind.Object)
+                {
+                    throw new InvalidOperationException($"The original JSON document to merge new content into must be a container type. Instead it is {root1.ValueKind}.");
+                }
+
+                if (root1.ValueKind != root2.ValueKind)
+                {
+                    return originalJson;
+                }
+
+                if (root1.ValueKind == JsonValueKind.Array)
+                {
+                    MergeArrays(jsonWriter, root2);
+                }
+                else
+                {
+                    MergeObjects(jsonWriter, root1, root2);
+                }
+            }
+
+            return Encoding.UTF8.GetString(outputBuffer.WrittenSpan);
+        }
+
+        /// <summary>
+        /// This helper methods tries to merge two Json Object recursively.
+        /// </summary>
+        private static void MergeObjects(Utf8JsonWriter jsonWriter, JsonElement root1, JsonElement root2)
+        {
+            jsonWriter.WriteStartObject();
+
+            // If a property exists in both documents, either:
+            // * Merge them, if the value kinds match (e.g. both are objects or arrays) recursively,
+            // * Completely override the value of the first with the one from the second, if the value kind mismatches (e.g. one is object, while the other is an array or string),
+            // * Or favor the value of the first, if the second one is null.
+            foreach (JsonProperty property in root1.EnumerateObject())
+            {
+                string propertyName = property.Name;
+
+                JsonValueKind newValueKind;
+
+                if (root2.TryGetProperty(propertyName, out JsonElement newValue) && (newValueKind = newValue.ValueKind) != JsonValueKind.Null)
+                {
+                    jsonWriter.WritePropertyName(propertyName);
+
+                    JsonElement originalValue = property.Value;
+                    JsonValueKind originalValueKind = originalValue.ValueKind;
+
+                    if (newValueKind == JsonValueKind.Object && originalValueKind == JsonValueKind.Object)
```

Comment:
>if (newValueKind == JsonValueKind.Object && originalValueKind == JsonValueKind.Object) [](http://example.com/codeflow?start=20&length=86)

So this gets pretty granular with respect to our config, yes? e.g. an override wouldn't generally replace the base config's "entity" object with the entire object present in the override file, but would only update/replace specific sections like "Source" "relationships" "rest" and "graphql"

also, permissions seem to be covered in the array case, that's good and reduces confusion from selective changes to field permissions/dbpolicies. However, this may require a documentation note, when the developer sets system role permissions in the base config file (anonymous and authenticated) but doesn't set them in the override file. Those two system roles would be overridden/replaced with what is in the override config.

-> would we just want to enable "an entity's config in the base file would just be replaced with the config from the override file?

Reply:
I can add examples in the comments, but in case of Json Arrays complete override will be performed.